src/udev/udev-node.c

```diff
@@ -212,26 +216,50 @@ static int link_update(sd_device *dev, const char *slink, bool add) {
         if (!add && unlink(filename) == 0)
                 (void) rmdir(dirname);
 
-        r = link_find_prioritized(dev, add, dirname, &target);
-        if (r < 0) {
-                log_device_debug(dev, "No reference left, removing '%s'", slink);
-                if (unlink(slink) == 0)
-                        (void) rmdir_parents(slink, "/");
-        } else
-                (void) node_symlink(dev, target, slink);
-
         if (add)
                 do {
                         _cleanup_close_ int fd = -1;
 
-                        r = mkdir_parents(filename, 0755);
+                        r = mkdir_parents_label(filename, 0755);
                         if (!IN_SET(r, 0, -ENOENT))
                                 break;
                         fd = open(filename, O_WRONLY|O_CREAT|O_CLOEXEC|O_TRUNC|O_NOFOLLOW, 0444);
                         if (fd < 0)
                                 r = -errno;
                 } while (r == -ENOENT);
 
+        for (int i = 0; i < LINK_UPDATE_MAX_RETRIES; i++) {
+                struct stat st1 = {}, st2 = {};
+
+                if (add && stat(dirname, &st1) < 0)
+                        return -errno;
+
+                r = link_find_prioritized(dev, add, dirname, &target);
+
+                if (add) {
+                        if (stat(dirname, &st2) < 0)
+                                return -errno;
+
+                        if (!stat_inode_unmodified(&st1, &st2))
+                                continue;
+                }
```

Comment:
this if block should be moved down, to the end of the loop iteration, i.e. after the unlink/node_symlink() call. i.e. if two workers run this code in parallel,  one one does the link_find_prioritized first, and gets rescheduled before it can issue the node_symlink() and another worker then gets shceduled and manages to go through both, then the first worker will update the symlink with out of date info once it is scheduled again. 

Hence: the mtime "bracket" should cover both parts: the determination of the priorized symlink and the creation of the symlink. i.e. the first stat() should be at the beginning of the loop iteration, the second at the very end of the loop iteration, and if mtimes still match the loop should be exited.

(oh, and we can optimize the stat thing here: move the first stat() before the loop, and then use the second stat()'s data as the first stat()'s data of the next loop iteration, if you follow what I mean? stat()s are cheap but not free, and in the contended case this would half the number of stat()s we need)

Reply:
> this if block should be moved down, to the end of the loop iteration, i.e. after the unlink/node_symlink() call. i.e. if two workers run this code in parallel, one one does the link_find_prioritized first, and gets rescheduled before it can issue the node_symlink() and another worker then gets shceduled and manages to go through both, then the first worker will update the symlink with out of date info once it is scheduled again.

This is handled by checking return value of `node_symlink()`. We can't rely only on mtime of symlink directory.

Reply:
the return code of node_symlink() has no effect in the current patch though: it just decides whether to use "continue" as last thing in the loop iteration or not, but that should not make any difference, see below.

why can't we rely on the mtime of the symlink dir?

Reply:
Because the algorithm as designed can still pick the wrong symlink, the race is still fundamentally there. If it does and it overwrites higher priority symlink (we didn't see that it is supposed to be higher priority when calling link_find_pioritized()) we should enter next loop iteration.

Reply:
I think the key here is that update of symlink dir and actual /dev links is done at different times.

Reply:
And update of db entry that we use to actually determine highest priority symlink happens at yet another time.

Reply:
it shouldn't matter if we pick up the wrong symlink momentarily, as long as we recover immediately. And we do: because we'll redo things immediately if the symlink db changed while we created the symlink.

As I said before: the symlink will not be totally off, it will just be "slightly" off, i.e. pointing to the next highest symlink or so, i.e. still to a device with the correct identifying data. And it will be corrected immediately.